### DISCUSSION TITLE
Switch the branch when using GitHub actions checkout

### DISCUSSION THREAD

**Discussion by huabin** [community/community#27102] [Actions]
<p>Hello,</p>
<p>I have some questions about GitHub actions <code>checkout</code>.</p>
<pre><code class="lang-auto">name: CI

# Controls when the action will run. 
on:
  # Triggers the workflow on push or pull request events but only for the master branch
  pull_request:
    branches: [ master ]

# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  # This workflow contains a single job called "deploy"
  deploy:
    # The type of runner that the job will run on
    runs-on: ubuntu-latest

    # Steps represent a sequence of tasks that will be executed as part of the job
    steps:
      # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it

      - name: Checkout
        uses: actions/checkout@v2

      - name: Add content to README.md
        run: echo 'New update' &gt;&gt; README.md
</code></pre>
<p>There’re two branches in my repo, the default one is <code>master</code> and another is <code>dev</code>, when I pull a request from <code>dev</code> to <code>master</code>.</p>
<ul>
<li>
<p>Which branch will be checkout here?</p>
</li>
<li>
<p>The following code works on which branch?</p>
<pre><code class="lang-auto">- name: Add content to README.md
  run: echo 'New update' &gt;&gt; README.md
</code></pre>
</li>
<li>
<p>How can I checkout the specified version?</p>
</li>
</ul>
<p>Thank you so much.</p>

**Comment by kingthorin** [comment, marked as answer]
<p>It’s covered in the docs:</p>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/actions/checkout" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/7570458/181101703-9fc0c086-6ae0-499e-a309-e9a44ce92439.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/4/f/4fc5c75a9b857a5e8c218c5e13bef6346270f8e1_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/4/f/4fc5c75a9b857a5e8c218c5e13bef6346270f8e1_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/4/f/4fc5c75a9b857a5e8c218c5e13bef6346270f8e1.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/4/f/4fc5c75a9b857a5e8c218c5e13bef6346270f8e1_2_10x10.png"></div>

<h3><a href="https://github.com/actions/checkout" target="_blank" rel="noopener nofollow ugc">actions/checkout</a></h3>


  <p><span class="label1">Action for checking out a repo. Contribute to actions/checkout development by creating an account on GitHub.</span></p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<blockquote>
<p>Refer <a href="https://help.github.com/en/articles/events-that-trigger-workflows" rel="noopener nofollow ugc">here</a> to learn which commit <code>$GITHUB_SHA</code> points to for different events.</p>
</blockquote>
<p>One of these two references should cover what you’re after:</p>
<ul>
<li><a href="https://github.com/actions/checkout#Checkout-a-different-branch" rel="noopener nofollow ugc">Checkout a different branch</a></li>
<li><a href="https://github.com/actions/checkout#Checkout-pull-request-HEAD-commit-instead-of-merge-commit" rel="noopener nofollow ugc">Checkout pull request HEAD commit instead of merge commit</a></li>
</ul>

**Comment by jsoref** [comment]
<aside class="quote no-group" data-username="hzb" data-post="1" data-topic="183055">
<div class="title">
<div class="quote-controls"></div>
 hzb:</div>
<blockquote>
<ul>
<li>Which branch will be checkout here?</li>
</ul>
</blockquote>
</aside>
<p>Technically (at least for a <code>dev</code> branch diverged from <code>master</code>), <em>neither</em>, as your event is <code>pull_request</code>, so you get the <code>/merge</code> of the <code>base</code> and the <code>request</code>.</p>
<p>If you want to be on <code>master</code> for the pull request, you could use <code>pull_request_target</code> instead (it’s a very different event, but, it does give you the destination, i.e. <code>master</code>).</p>

**Comment by huabin** [comment]
<p>I see, thank you so much.</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="jsoref" data-post="3" data-topic="183055">
<div class="title">
<div class="quote-controls"></div>
 jsoref:</div>
<blockquote>
<p>If you want to be on <code>master</code> for the pull request, you could use <code>pull_request_target</code> instead (it’s a very different event, but, it does give you the destination, i.e. <code>master</code> ).</p>
</blockquote>
</aside>
<p>That is dangerous, because <code>pull_request_target</code> workflows run with additional privileges (see also: <a href="https://securitylab.github.com/research/github-actions-preventing-pwn-requests/" class="inline-onebox">Keeping your GitHub Actions and workflows secure: Preventing pwn requests | GitHub Security Lab</a>).</p>
<p>If all you want is to check out a specific branch, you can use the <code>ref</code> parameter to <code>actions/checkout</code>, as described in the documentation <a class="mention" href="/kingthorin">@kingthorin</a> linked already:</p>
<pre><code class="lang-yaml">- uses: actions/checkout@v2
  with:
    ref: my-branch
</code></pre>

**Comment by jsoref** [comment]
<p>The example clearly indicated intent to perform an append to an existing file.</p>
<p>Sure, it’s dangerous. But it isn’t going to work if you use <code>pull_request</code>.</p>
<p>I did indicate that the event was very different. I expect that people read the docs when given advice to consider something new.</p>
<p>And the other answer was already given. Repeating it would not have been a good use of space.</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="jsoref" data-post="6" data-topic="183055">
<div class="title">
<div class="quote-controls"></div>
 jsoref:</div>
<blockquote>
<p>The example clearly indicated intent to perform an append to an existing file.</p>
<p>Sure, it’s dangerous. But it isn’t going to work if you use <code>pull_request</code> .</p>
</blockquote>
</aside>
<p>True, assuming the file should be pushed back to the repository. Unless PR are only from within the repository.</p>
<aside class="quote no-group" data-username="jsoref" data-post="6" data-topic="183055">
<div class="title">
<div class="quote-controls"></div>
 jsoref:</div>
<blockquote>
<p>I did indicate that the event was very different. I expect that people read the docs when given advice to consider something new.</p>
</blockquote>
</aside>
<p>Unfortunately in my experience that assumption is wrong more often than not. And either way I think a warning is appropriate when recommending something with potentially dangerous side-effects.</p>

**Comment by jsoref** [comment]
<p>Fair enough. Will you agree we both made foolish assumptions here and leave this conversation alone?</p>
